SNS Asset Canister page (#1613)
dfinity/portal: docs/developer-docs/integrations/sns/managing-sns/sns-asset-canister.md

# SNS asset canisters

## Overview

The [asset canister](https://github.com/dfinity/sdk/tree/master/src/canisters/frontend/ic-frontend-canister) provides users with a way to store and retrieve static assets from a canister deployed on the IC. Generally, asset canisters are used to serve HTML, CSS, or JavaScript assets, which are typically part of a dapp's frontend. For this reason, the asset canister is also referred to as the frontend canister. For purposes of this guide, we'll refer to it as the asset canister. 

In the context of the SNS, a dapp's associated asset canister serves the frontend assets related to dapp and may also include a frontend to the SNS DAO, e.g., through which users can vote on governance proposals.  

> The contents of the asset canister must be configured prior to the launch of the SNS, and any changes afterwards must be made by a principal with the permission to do so. The SNS is the only one that can update the list of approved principals that have this permission, and any additions to this list approved on via an SNS proposal. This configuration assures that changes to the asset canister are only made by authorized principals that the SNS has agreed on through proposals. These changes are referred to as 'upgrades' to the asset canister in the remainder of this document. 

Reviewer: ```suggestion
The contents of the asset canister must be configured prior to the launch of the SNS, and any changes afterwards must be made by a principal with the permission to do so. The SNS is the only one that can update the list of approved principals that have this permission, and any changes to this list must be approved on via an SNS proposal. This configuration assures that changes to the asset canister are only made by authorized principals that the SNS has agreed on through proposals. These changes are referred to as 'upgrades' to the asset canister in the remainder of this document. 
```

> The contents of the asset canister must be configured prior to the launch of the SNS, and any changes afterwards must be made by a principal with the permission to do so. The SNS is the only one that can update the list of approved principals that have this permission, and any additions to this list approved on via an SNS proposal. This configuration assures that changes to the asset canister are only made by authorized principals that the SNS has agreed on through proposals. These changes are referred to as 'upgrades' to the asset canister in the remainder of this document. 

Reviewer: So I think this is now not true anymore. Sorry for the confusion! It sounds like the old text was right here: any change needs to approved by the SNS

When a dapp's control is handed over to an SNS, this is also true for the associated asset canister.

This section is relevant if your project contains an asset canister and describes how you can test handing over control of an asset canister to an SNS.

> 

Reviewer: Would it make sense to have here a high level introduction how all of this works? 
I imagine the following content:
* what are the roles that can be set in the asset canister and what can these roles do?
* In a non-DAO setting who would have these roles? 
* Once asset canister is handed over to SNS DAO, who has which roles and can do what?

> ## Deploying an asset canister

Reviewer 2: Instead of going through all this trouble with `curl` and a custom canister, people can also simply declare a canister as type `assets` and dfx will do all of the below automatically

> 

Reviewer: I would maybe start here saying that we assume that the asset cansiter is deployed before it is handed over to the SNS. We explain how one can do so locally for testing and how one can do so in production.

Then have two subsections as now.

```
curl -L "https://github.com/dfinity/sdk/raw/${DX_COMMIT}/src/distributed/assetstorage.wasm.gz" -o assets.wasm.gz
curl -L "https://raw.githubusercontent.com/dfinity/sdk/${DX_COMMIT}/src/distributed/assetstorage.did" -o candid/assets.did
```

Next, you will need to configure your `dfx.json` file to specify your asset canister. An example `dfx.json` file can be found below:

```
{
  "canisters": {
    "nns-dapp": {
      "type": "custom",
      "candid": "",
      "wasm": "nns-dapp.wasm"
    },
    "internet_identity": {
      "type": "custom",
      "candid": "",
      "wasm": "internet_identity.wasm"
    },
    "sns_aggregator": {
      "type": "custom",
      "candid": "candid/sns_aggregator.did",
      "wasm": "sns_aggregator.wasm"
    },
    "test": {
      "candid": "candid/test.did",
      "type": "rust",
      "package": "test"
    },
    "assets": {
      "build": "",
      "candid": "candid/assets.did",
      "type": "custom",
      "wasm": "./assets.wasm.gz"
    }
  }
}
```

### Deploying locally

To deploy your asset canister locally for testing purposes, the following command can be used:

```
dfx deploy assets --network "local" --no-wallet
```

> Once the asset canister has been deployed, the SNS can be launched. You can learn more about launching an SNS [here](../launch-sns/launch-sns.md).

Reviewer: I am not sure I follow here. This talks about deploying the asset canister locally and then we say "and SNS can be launched".
Do we mean an SNS on a testnet?
Maybe the whole section could be called "deploying locally for testing" and then we can say here "once..., the SNS can also be launched locally to test it" (and then link to the test sns page rather than the launch page)?

### Deploying on the mainnet

> 

Reviewer: Should we maybe start here with "the asset cansiter should be deployed before the dapp is handed over to an SNS. To deploy the asset canister on mainnet... "

:::info 

> To deploy to the mainnet network, you will need a wallet that contains cycles. For more information on cycles wallets, please see [here](../../../setup/cycles/cycles-wallet.md).

Reviewer: ```suggestion
To deploy to the mainnet, you will need a wallet that contains cycles. For more information on cycles wallets, please see [here](../../../setup/cycles/cycles-wallet.md).
```

(I though the "net" in _mainnet_ is already short for "network", but maybe there is a convention that I am not aware of)

:::

To deploy your asset canister to the mainnet, the following command can be used:

```
dfx deploy assets --network "ic" --wallet <principal>
```

> Once the asset canister has been deployed, the SNS can be launched. You can learn more about launching an SNS [here](../launch-sns/launch-sns.md)

Reviewer: I think we should add here that during the SNS launch, the control of the asset canister has to be handed over to the SNS root. Moreover, the permissions must be set correctly in this step. 
Maybe we could refer here to the steps in the launch process that @dprats is working on?

> 

Reviewer: 1. In the end of the deployment, would it make sense to link to docs that explain how the assets can be upgraded as long as the asset cansiter is deployed but not yet handed over to an SNS?

2. Between the "deployment" and the "submitting an SNS proposal"..  I would  maybe have a section saying what a dev has to do with the asset canister during the SNS launch (hand it over, set the roles correctly so that afterwards everything works, register the correct generic proposals).

## Submitting an SNS proposal and upgrading an asset canister

> 

Reviewer: Maybe we can add a sentence here that this is the flow how to update assets once the asset cansiter is already under SNS control?

> - #### Step 1: First, upgrade the asset canister (by proposal, as with any other canister) to the asset canister wasm bundled with dfx 0.14.1, available [here](https://github.com/dfinity/sdk/blob/release-0.14.1/src/distributed/assetstorage.wasm.gz).

Reviewer: I assume this is only necessary now (and not in later versions), right @sesi200 ? 
If so, maybe we can first say "prerequisites are that you are on this version or newer, otherwise as step 1, upgrade..)

> - #### Step 2: After upgrading the asset canister by proposal, have someone with `Prepare` permission run:

Reviewer: should we maybe have a section in the beginning introducing all the different roles and what they mean?

```
dfx deploy <frontend canister name> --network ic --compute-evidence
```

The computed evidence should match the evidence from step 2.

- #### Step 5: Submit a new proposal to commit the batch, using the APIs below:

Use these asset canister APIs in the proposal:

```
   type CommitProposedBatchArguments = record {
   batch_id: BatchId;
   evidence: blob;
  };
  type ValidationResult = variant { Ok : text; Err : text };

    validate_commit_proposed_batch: (CommitProposedBatchArguments) -> (ValidationResult);
  commit_proposed_batch: (CommitProposedBatchArguments) -> ();

```

An example can be found [here](https://github.com/dfinity/sdk/blob/master/e2e/tests-dfx/assetscanister.bash#L133).

If the proposal is rejected, the preparer should use this new asset canister API:

```
  type DeleteBatchArguments = record {
    batch_id: BatchId;
  };
  delete_batch: (DeleteBatchArguments) -> ();
```

## Asset canister examples

An example of an SNS asset canister is canister `sqbzf-5aaaa-aaaam-aavya-cai`, which is an asset canister part of the [Dragginz Dapp SNS](https://dashboard.internetcomputer.org/canister/sqbzf-5aaaa-aaaam-aavya-cai).

## SNS GenericNervousSystemFunctions

> To submit a new `AddGenericNervouSystemFunction` SNS Proposal to support the `commit_proposed_batch` API, the target canister id should be the asset canister (that is upgraded in upgrade steps) and the target function is `commit_proposed_batch`. The validate function should be `validate_commit_proposed_batch`. 

Reviewer: I think this needs to happen before the above upgrade can work. Therefore I would maybe restructure this sligthly so that it is in order of the actions that a dev has to do. WDYT?

# SNS asset canisters

## Overview

> The [asset canister](https://github.com/dfinity/sdk/tree/master/src/canisters/frontend/ic-frontend-canister) provides users with a way to store and retrieve static assets from a canister deployed on the IC. Generally, asset canisters are used to serve HTML, CSS, or JavaScript assets, which are typically part of an application's frontend. For this reason, the asset canister is also referred to as the frontend canister. For purposes of this guide, we'll refer to it as the asset canister. 

Reviewer: ```suggestion
The [asset canister](https://github.com/dfinity/sdk/tree/master/src/canisters/frontend/ic-frontend-canister) provides users with a way to store and retrieve static assets from a canister deployed on the IC. Generally, asset canisters are used to serve HTML, CSS, or JavaScript assets, which are typically part of an dapp's frontend. For this reason, the asset canister is also referred to as the frontend canister. For purposes of this guide, we'll refer to it as the asset canister. 
```

> The asset canister of an SNS serves the frontend assets related to the SNS. The contents of the asset canister must be configured prior to the launch of the SNS, and any changes afterwards must be approved on via an SNS proposal. These changes are referred to as 'upgrades' to the asset canister in the remainder of this document. 

Reviewer: I am not sure I agree with this as the asset canister can also just be the frontend to the dapp that is governed by the SNS DAO. Let me try to make an edit proposal... 
```suggestion
In the context of the SNS, a dapp's associated asset canister serves the frontend assets related to dapp and may also include a frontend to the SNS DAO, e.g., through which users can vote on governance proposals. 
When a dapp's control is handed over to an SNS, this is also true for the associated asset canister. 
After handing over control to the SNS, asset canister changes, referred to as 'upgrades' must be approved by the SNS DAO.
```

This section is relevant if your project contains an asset canister and describes how you can test handing over control of an asset canister to an SNS.

:::info

> Note that you can interleave the steps for the asset canister from this section and the steps for the test canister (and also your own canister) within the SNS lifecycle. We list additional constraints in between the steps below.

Reviewer: I am not sure if it is clear how this page relates to all other pages..
Do we describe testing here? or what to do in production? or both?

:::

## Deploying an asset canister

To deploy an asset canister, the `./deploy_assets.sh` script found in the [SNS testing repo](https://github.com/dfinity/sns-testing) can be used.

> Then, the SNS can be deployed using the `./deploy_sns.sh <config-path>` script. After the SNS has been deployed, the following steps can be used to prepare the SNS canister for being handed off to the SNS:

Reviewer: where is this script? could it be that this text was written as part of a README in a repository where this script was?
I wonder if it is clear here what this refers to.

```
dfx canister --network ${NETWORK} call assets list_permitted '(record {permission = variant {<permission>}})'
```

or by running the script:

```
./get_permission_assets.sh.
```

- #### Step 4: Make sure that you can still commit assets by running the script:

```
./commit_assets.sh <path> <content>
```

where `<path>` is the HTTP path of the asset, e.g., `/myasset.txt` and `<content>` is the ASCII-encoded content of the asset.

- #### Step 5: Testing all possible SNS launch scenarios includes testing a failed swap (e.g., if not enough funds have been raised) where the control of the asset canister is given back to your principal.

You should run the following step after `finalize_sns_swap.sh` for an unsuccessful swap:

   - After a failed swap, you can run the script:
   
   ```
   ./take_ownership_assets.sh
   ```
   
   to reset the permissions of the asset canister back to only your principal having the Commit permission (in particular, with the SNS governance having no permission anymore).

> ## Submitting a proposal 

Reviewer 3: can you specify if its an SNS proposal or an NNS proposal (I myself do not know)

Imagine a scenario where you'd like to upload a frontend file to the SNS's asset canister. To do this, the governance canister must call the asset canister to upload the file. 

- #### Step 1: To get the governance canister to call the asset canister, we first make the proposal to `AddGenericNervousSystemFunction`, which is like adding a new proposal type that uploads files to the asset canister.

> - #### Step 2: Next, we can make an ExecuteGenericNervousSystemFunction proposal that will make the governance-canister call the asset-canister’s store method with a payload of the frontend asset Candid parameter for the ‘store’ method.

Reviewer 3: `ExecuteGenericNervousSystemFunction`  with `backticks` perhaps?

      "candid": "candid/test.did",
      "type": "rust",
      "package": "test"
    },
    "assets": {
      "build": "",
      "candid": "candid/assets.did",
      "type": "custom",
      "wasm": "./assets.wasm.gz"
    }
  }
}
```

Alternatively, any canister in the `dfx.json` file can be set as `"type": "assets"`, and dfx will automatically perform the above steps. 

### Deploying locally for testing

To deploy your asset canister locally for testing purposes, the following command can be used:

```
dfx deploy assets --network "local" --no-wallet
```

The asset canister should be deployed and have permissions configured before the dapp is handed over to an SNS. 

### Deploying on the mainnet

:::info 
To deploy to the mainnet, you will need a wallet that contains cycles. For more information on cycles wallets, please see [here](../../../setup/cycles/cycles-wallet.md).
:::

To deploy your asset canister to the mainnet, the following command can be used:

```
dfx deploy assets --network "ic" --wallet <principal>
```

The asset canister should be deployed and have permissions configured before the dapp is handed over to an SNS. 

## Configuring an asset canister's permissions

> When configuring an asset canister, a set of permissions that contains a whitelist of principals must be created. This whitelist details who is allowed submit proposals that update assets in the asset canister. This whitelist must be configured prior to the SNS launch. Principals that are allowed to submit proposals are given the `Prepare` permission.

Author: You are correct Lara, per @sesi200 on Slack:

"You need `Prepare` to make a batch of changes to the asset canister and ‘lock’ them. To make the SNS proposal, you don’t need any special permissions - everyone can do that. Reason is that if you create a bunch of bogus SNS proposals to commit changes you’ll get charged the reject fee every time. No need to put that behind a permission barrier."

I'll revise and edit to reflect this.

```

### Listing permissions

To list permissions for an asset canister, the following command can be used:

```
dfx canister call --network ic <canister-id> list_permitted '(record {permission = variant {<Permission>}})'
```

For example, to list all principals with the `Commit` permission:

```
dfx canister call --network ic oa7fk-maaaa-aaaam-abgka-cai list_permitted '(record {permission = variant {Commit}})'
```

To list all principals with the `Prepare` permission:

```
dfx canister call --network ic oa7fk-maaaa-aaaam-abgka-cai list_permitted '(record {permission = variant {Commit}})'
```

Once the asset canister has been deployed and permissions have been configured, the SNS can be launched. You can learn more about launching an SNS [here](../launch-sns/launch-sns.md)

## SNS GenericNervousSystemFunctions

To submit a new `AddGenericNervousSystemFunction` SNS Proposal to support the `commit_proposed_batch` API, the target canister id should be the asset canister (that is upgraded in the upgrade steps below) and the target function is `commit_proposed_batch`. The validate function should be `validate_commit_proposed_batch`. 

> To submit an `ExecuteNervousSystemFunction` SNS Proposal with the output from `dfx deploy <frontend canister name> --network ic --by-proposal` see upgrade steps below.

Reviewer: could add the info that this part is now defining how to execute the proposal that is now registered
